### PR TITLE
[SYCLomatic] Refine the dpct::sparse::sparse_matrix_desc

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 1]
```diff
@@ -667,50 +647,39 @@ class sparse_matrix_desc {
   std::int64_t get_row_num() const noexcept { return _row_num; }
 
 private:
+  std::function<void(void *)> _shadow_row_ptr_deleter = [](void *ptr) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: static

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 2]
```diff
@@ -667,50 +647,39 @@ class sparse_matrix_desc {
...
+    if (ptr)
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: remove if

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 3]
```diff
@@ -667,50 +647,39 @@ class sparse_matrix_desc {
...
+  template <typename index_t, typename value_t> void set_data() {
+    void *row_ptr = nullptr;
+    if (_shadow_row_ptr) {
+      row_ptr = _shadow_row_ptr.get();
+    } else if (_row_ptr) {
+      row_ptr = _row_ptr;
+    } else {
+      row_ptr = dpct::dpct_malloc(sizeof(index_t) * (_row_num + 1),
+                                  get_default_queue());
+      _shadow_row_ptr = std::unique_ptr<void, std::function<void(void *)>>(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: use reset member function

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 4]
```diff
@@ -785,17 +753,34 @@ class sparse_matrix_desc {
...
+
+  static constexpr std::array<size_t, 7> _data_variable_size_array = {
+      sizeof(sycl::buffer<std::int32_t>),
+      sizeof(sycl::buffer<std::int64_t>),
+      sizeof(sycl::buffer<float>),
+      sizeof(sycl::buffer<double>),
+      sizeof(sycl::buffer<std::complex<float>>),
+      sizeof(sycl::buffer<std::complex<double>>),
+      sizeof(void *)};
+  static constexpr size_t _max_data_variable_size = *std::max_element(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: use std::max

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 5]
```diff
@@ -785,17 +753,34 @@ class sparse_matrix_desc {
...
+      _data_variable_size_array.begin(), _data_variable_size_array.end());
+  std::variant<std::array<std::byte, _max_data_variable_size>,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: typedef

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 6]
```diff
@@ -785,17 +753,34 @@ class sparse_matrix_desc {
...
+               sycl::buffer<std::int32_t>, sycl::buffer<std::int64_t>,
+               std::int32_t *, std::int64_t *>
+      _data_row_ptr;
+  std::variant<std::array<std::byte, _max_data_variable_size>,
+               sycl::buffer<std::int32_t>, sycl::buffer<std::int64_t>,
+               std::int32_t *, std::int64_t *>
+      _data_col_ind;
+  std::variant<std::array<std::byte, _max_data_variable_size>,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: typedef

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 2 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 3 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 5 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 6 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

**Review by ziranzha — APPROVED** [review]

**Review by tomflinda — APPROVED** [review]
LGTM

**Pull request merged**